[DOCS] Update subcurrency example in introductory section (#6567)
ethereum/solidity: docs/introduction-to-smart-contracts.rst

Subcurrency Example
===================

The following contract will implement the simplest form of a
The following contract implements the simplest form of a
cryptocurrency. It is possible to generate coins out of thin air, but
only the person that created the contract will be able to do that (it is easy
only the person that created the contract is able to do that (it is possible
to implement a different issuance scheme).
Furthermore, anyone can send coins to each other without a need for
registering with username and password — all you need is an Ethereum keypair.

registering with username and password, all you need is an Ethereum keypair.

::

    pragma solidity >=0.5.0 <0.7.0;

    contract Coin {

>         // The keyword "public" makes those variables

Reviewer: Don't you think these inline comments make it a bit more readable? With inline comments, you do not always have to flick back and forth and it is always clear (at least almost) what the text refers to.

Author: Personally I found it confusing that you see everything (without knowing what any of it is) and then you get the explanation, which for a beginner is a little confusing. Personally I had to go back and forth a lot with the current version to understand what lines meant, and re-read several times.

I have two ideas.

1. Add back the comments, but keep some of the further explanations after the code, offering people two ways to understand what the code does.
2. Change the order. Start with a blank contract, and add and explain each section piece by piece and then present the final code at the end. This is a more traditional way of structuring a getting started guide.

Reviewer: I agree with you, but I think removing the comments exactly causes the problem that you see something without explanation ;)

Having said that, the exact comments might perhaps not be too helpful. Reading about "light clients" or things being "easily accessible from outside" does not really help people understand what this is about. We already know what state variables are, so this is good, but perhaps at least the comment on the constructor should stay? Perhaps you could also find better comments that explain in a single sentence what we are seeing here?

Subcurrency Example
===================

The following contract implements the simplest form of a
cryptocurrency. The contract allows only its creator to create new coins (different issuance scheme are possible).
Anyone can send coins to each other without a need for
registering with a username and password, all you need is an Ethereum keypair.

::

    pragma solidity >=0.5.0 <0.7.0;

    contract Coin {
        // The keyword "public" makes variables
        // accessible from other contracts
        address public minter;
        mapping (address => uint) public balances;

        // Events allow clients to react to specific
        // contract changes you declare
        event Sent(address from, address to, uint amount);

        // Constructor code is only run when the contract
        // is created
        constructor() public {
            minter = msg.sender;
        }

        // Sends an amount of newly created coins

>         // from the contract creator to an address

Reviewer: The contract creator does not really own them, so they are not sent from the contract creator.

Reviewer: Creates an amount of coins and assigns them to a specified address?

Reviewer: (can only be called by the contract creator)

This contract introduces some new concepts, let us go through them one by one.

> The line ``address public minter;`` declares a state variable of type address

Reviewer: Please do not change the line breaks here.

Author: This was somewhat intentional, to break up the spacing and thus flow of a lot of explanation for someone new to Solidity, breaking each concept into a new paragraph, making it easier to scan through them one by one. White space helps with easing this flow. Unless I am misunderstanding what you mean?

Reviewer: I mean the line breaks in the source code, not in the rendered output - it is hard to review this way.

This contract introduces some new concepts, let us go through them one by one.

The line ``address public minter;`` declares a state variable of type :ref:`address<address>`.
The ``address`` type is a 160-bit value that does not allow any arithmetic operations.

> It is suitable for storing addresses of contracts, or of keypairs belonging to external persons.

Reviewer: keypairs are not stored, it is only a hash of the public part of a keypair that is stored.

Mappings can be seen as `hash tables <https://en.wikipedia.org/wiki/Hash_table>`_ which are
virtually initialised such that every possible key exists from the start and is mapped to a
value whose byte-representation is all zeros. However, it is neither possible to obtain a list of all keys of
a mapping, nor a list of all values. Keep in mind what you
added to the mapping or use it in a context where this is not needed. Or
even better, keep a list, or use a more suitable data type.

The :ref:`getter function<getter-functions>` created by the ``public`` keyword
is more complex in the case of a mapping. It looks like the
following::

    function balances(address _account) external view returns (uint) {
        return balances[_account];
    }

You can use this function to query the balance of a single account.

.. index:: event

The line ``event Sent(address from, address to, uint amount);`` declares
an :ref:`"event" <events>`, which is emitted in the last line of the function
``send``. Light Ethereum clients such as web applications can
listen for these events emitted on the blockchain without much

> gas cost. As soon as it is emitted, the listener receives the

Reviewer: ```suggestion
cost. As soon as it is emitted, the listener receives the
```

This contract introduces some new concepts, let us go through them one by one.

The line ``address public minter;`` declares a state variable of type :ref:`address<address>`.
The ``address`` type is a 160-bit value that does not allow any arithmetic operations.

> It is suitable for storing addresses of contracts, or a hash of the public half of a keypair belonging to external persons.

Reviewer: Perhaps we should just go with `addresses of contracts and external accounts`? Maybe the blockchain basics can be linked to explain "external account"?

The keyword ``public`` automatically generates a function that allows you to access the current value of the state
variable from outside of the contract. Without this keyword, other contracts have no way to access the variable.
The code of the function generated by the compiler is equivalent
to the following (ignore ``external`` and ``view`` for now)::

    function minter() external view returns (address) { return minter; }

You could add a function like the above yourself, but you would have a function and state variable with the same name.
You do not need to do this, the compiler figures it out for you.

.. index:: mapping

The next line, ``mapping (address => uint) public balances;`` also
creates a public state variable, but it is a more complex datatype.
The :ref:`mapping <mapping-types>` type maps addresses to :ref:`unsigned integers <integers>`.

Mappings can be seen as `hash tables <https://en.wikipedia.org/wiki/Hash_table>`_ which are
virtually initialised such that every possible key exists from the start and is mapped to a
value whose byte-representation is all zeros. However, it is neither possible to obtain a list of all keys of

> a mapping, nor a list of all values. Keep in mind what you

Reviewer: ```suggestion
a mapping, nor a list of all values. Either record what you
```
(the developer remembering it does not really help...)

added to the mapping or use it in a context where this is not needed. Or

> even better, keep a list, or use a more suitable data type.

Reviewer: And then this can be deleted

Author: Which part? I feel like at least a part of this explanation is still useful.

Reviewer: The sentence "Or even better, keep a list, ..."

Reviewer: Actually it is fine as it is.

The :ref:`getter function<getter-functions>` created by the ``public`` keyword
is more complex in the case of a mapping. It looks like the
following::

    function balances(address _account) external view returns (uint) {
        return balances[_account];
    }

You can use this function to query the balance of a single account.

.. index:: event

The line ``event Sent(address from, address to, uint amount);`` declares
an :ref:`"event" <events>`, which is emitted in the last line of the function

> ``send``. Light Ethereum clients such as web applications can

Reviewer: Over the years, it turned out that even non-light ethereum clients can not properly function without events, so maybe just change this into "Ethereum applications"?

        }
    })

> .. note::

Reviewer: I don't think this should be made into a note-block.

    The user interface calls the automatically generated ``balances`` function from above.

.. index:: coin

The :ref:`constructor<constructor>` is a special function run during the creation of the contract and
cannot be called afterwards. It permanently stores the address of the person creating the
contract. The ``msg`` variable (together with ``tx`` and ``block``) is a
:ref:`special global variable <special-variables-functions>` that
contains properties which allow access to the blockchain. ``msg.sender`` is
always the address where the current (external) function call came from.

The functions that make up the contract, and that users and contracts can call are ``mint`` and ``send``.

> The ``mint`` function sends an amount of newly created coins from the account that created the contract to another address.

Reviewer: Also: not from the creator

> The :ref:`require <assert-and-require>` function defines conditions that throw an exception if not met, and reverts all changes. In

Reviewer: I'm not sure we should use the word exception here. Also: The conditions do not throw the exception.

Author: @chriseth I took the wording from here, so if it needs to change, I am not sure exactly what to, and we should change both.

Reviewer: We used `exception` in the past but are moving towards `revert` or `failing assertion` now.

Author: OK, I'll change here and make some future PRs to change this elsewhere.

        }
    })

.. index:: coin

The :ref:`constructor<constructor>` is a special function run during the creation of the contract and

> cannot be called afterwards. It permanently stores the address of the person creating the

Reviewer: ```suggestion
cannot be called afterwards. In this case, it permanently stores the address of the person creating the
```
(to make it clear that this is not automatically the case for all contracts)

contract. The ``msg`` variable (together with ``tx`` and ``block``) is a
:ref:`special global variable <special-variables-functions>` that
contains properties which allow access to the blockchain. ``msg.sender`` is
always the address where the current (external) function call came from.

The functions that make up the contract, and that users and contracts can call are ``mint`` and ``send``.

The ``mint`` function sends an amount of newly created coins to another address.

> The :ref:`require <assert-and-require>` function defines conditions that throw an exception if not met, and reverts all changes. In

Reviewer: Also, the function does not define the condition, the function call does it. But maybe that is pedantic...

this example, ``require(msg.sender == minter);`` ensures that only the creator of the contract can call ``mint``,

> and ``require(amount < 1e60);`` ensures a maximum amount of tokens, which could cause overflow errors in the future.

Reviewer: "The absence of which could cause..."?

Reviewer: Or just "without it, we might run into overflow problems"
